napari/layers/transforms.py

```diff
@@ -0,0 +1,173 @@
+        A function converting an NxD array of coordinates to NxD'.
+    """
+
+    def __init__(self, func=tz.identity, inverse=None, name=None):
+        self.func = func
+        self._inverse_func = inverse
+        self.name = name
+
+        if func is tz.identity:
+            self._inverse_func = tz.identity
+
+    def __call__(self, coords):
+        """Transform input coordinates to output."""
+        return self.func(coords)
+
+    def set_slice(self, axes: Sequence[int]) -> Transform:
+        """Return a transform subset to the visible dimensions."""
+        raise NotImplementedError('Cannot subset arbitrary transforms.')
+
+    def set_pad(self, axes: Sequence[int]) -> Transform:
+        """Return a transform with added axes for non-visible dimensions."""
+        raise NotImplementedError('Cannot subset arbitrary transforms.')
+
+    @property
+    def inverse(self):
+        if self._inverse_func is not None:
+            return Transform(self._inverse_func, self.func)
+        else:
+            raise ValueError('Inverse function was not provided.')
+
+
+# class TransformChain(ListModel, Transform):
+#     def __init__(self, transforms=[]):
+#         super().__init__(
+#             basetype=Transform,
+#             iterable=transforms,
+#             lookup={str: lambda q, e: q == e.name},
+#         )
+#
+#     def __call__(self, coords):
+#         return tz.pipe(coords, *self)
+#
+#     def __newlike__(self, iterable):
+#         return ListModel(self._basetype, iterable, self._lookup)
+#
+#     def set_slice(self, axes: Sequence[int]) -> TransformChain:
+#         return TransformChain([tf.set_slice(axes) for tf in self])
+
+
+class STTransform(Transform):
+    """n-dimensional scale and translation (shift) class.
+
+    Scaling is always applied before translation.
+
+    An empty translation vector implies no translation, and an empty scale
+    implies no scaling.
+
+    Translation is broadcast to 0 in leading dimensions, so that, for example,
+    a translation of [4, 18, 34] in 3D can be used as a translation of
+    [0, 4, 18, 34] in 4D without modification.
+
+    Translation is broadcast to 1 in leading dimensions, so that, for example,
+    a scale of [4, 18, 34] in 3D can be used as a scale of
+    [1, 4, 18, 34] in 4D without modification.
+    """
+
+    def __init__(self, scale=(1.0,), translate=(0.0,), name=None):
+        super().__init__(name=name)
+        self.scale = np.array(scale)
+        self.translate = np.array(translate)
+
+    def __call__(self, coords):
+        coords = np.atleast_2d(coords)
+        scale = np.concatenate(
+            ([1.0] * (coords.shape[1] - len(self.scale)), self.scale)
+        )
+        translate = np.concatenate(
+            ([0.0] * (coords.shape[1] - len(self.translate)), self.translate)
+        )
+        return np.squeeze(scale * coords + translate)
+
+    @property
+    def inverse(self):
+        return STTransform(1 / self.scale, -self.translate)
```

Comment:
So this turns out to be incorrect, because the inverse of doing a scaling followed by a translation, is a translation followed by a scaling, which cannot be represented by this class. Another reason why I liked implementing this as a chain. ;)

```python
In [2]: import numpy as np                                                         
In [3]: arr = np.array([[0.5, 0, 45], [0, 0.5, 89], [0, 0, 1]])                    
In [4]: np.linalg.inv(arr)                                                         
Out[4]: 
array([[   2.,    0.,  -90.],
       [   0.,    2., -178.],
       [   0.,    0.,    1.]])
```

You can see that if you want to implement it this way, you will have to divide translate by scale also.

Reply:
yup absolutely, good catch - thanks - now the error is caught by a test too :-) Dividing by the translate by the scale is totally fine though to make the inverse, you have to do a similar option for an affine more generally and I think independent of the chain stuff. I'll respond in the main thread about why i like scale and translate going together, but this is at least now fixed and tested